### PR TITLE
feat: auto retriever

### PR DESCRIPTION
**Pull request by Wendong-Fan**
## Description

This this a sub PR of https://github.com/camel-ai/camel/pull/410 only containing auto retriever module based on vector retriever
## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/retrievers/auto_retriever.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+            "Milvus", "Qdrant".
+        embedding_model (BaseEmbedding): Model used for embedding queries and
+            documents. Defaults to `OpenAIEmbedding()`.
+    """
+
+    def __init__(
+        self,
+        url_and_api_key: Optional[Tuple[str, str]] = None,
+        vector_storage_local_path: Optional[str] = None,
+        storage_type: Optional[str] = None,
```
Comment (zechengz) [inline camel/retrievers/auto_retriever.py:55]: Make this as Enum type?

File: camel/retrievers/auto_retriever.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                collection_name=collection_name,
+                path=self.vector_storage_local_path,
+                url_and_api_key=self.url_and_api_key)
+
+        raise ValueError(
+            f"Unsupported vector storage type: {self.storage_type}, supported"
+            "type: 'milvus', 'qdrant'")
+
+    def _get_file_modified_date(self, content_input_path: str) -> str:
+        r"""Retrieves the last modified date and time of a given file. This
```
Comment (zechengz) [inline camel/retrievers/auto_retriever.py:99]: nit
```suggestion
        r"""Retrieves the last modified time of a given file. This
```

File: camel/retrievers/auto_retriever.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                including similarity score, content path and metadata.
+
+        Raises:
+            ValueError: If there's an vector storage existing with content
+                name in the vector path but the payload is None.
+            RuntimeError: If any errors occur during the retrieve process.
+        """
+        vr = VectorRetriever()
+        content_input_paths = [content_input_paths] if isinstance(
+            content_input_paths, str) else content_input_paths
```
Comment (zechengz) [inline camel/retrievers/auto_retriever.py:148]: nit: raise an error if `content_input_paths` is empty?

File: camel/retrievers/auto_retriever.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+        for content_input_path in content_input_paths:
+            # Check path type
+            parsed_url = urlparse(content_input_path)
+            is_url = all([parsed_url.scheme, parsed_url.netloc])
+
+            # Convert given path into a collection name, ensuring it only
+            # contains numbers, letters, and underscores
+            if is_url:
+                # For URLs, replace https://, /, and any characters not
```
Comment (zechengz) [inline camel/retrievers/auto_retriever.py:161]: Here we will remove `https://` by replacing it as an empty string right?

File: camel/retrievers/auto_retriever.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+                    content_input_path.replace("https://", ""))
+            else:
+                # For file paths, get the stem and replace spaces with _, also
+                # ensuring only allowed characters are present
+                collection_name = re.sub(r'[^0-9a-zA-Z]+', '_',
+                                         Path(content_input_path).stem)
+
+            # Ensure the collection name does not start or end with an
+            # underscore
+            collection_name = collection_name.strip("_")
```
Comment (zechengz) [inline camel/retrievers/auto_retriever.py:174]: Do we need to limit the max length of the collection name? Is it possible that collection name may be very long and make the vector storage initialization fail?

File: camel/retrievers/auto_retriever.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+        Returns:
+            str: The last modified date and time of the file.
+        """
+        mod_time = os.path.getmtime(content_input_path)
+        readable_mod_time = datetime.datetime.fromtimestamp(
+            mod_time).isoformat(timespec='seconds')
+        return readable_mod_time
+
+    def run_vector_retriever(
```
Comment (zechengz) [inline camel/retrievers/auto_retriever.py:115]: This function is too long. Can you refactor a little bit?

**Review by zechengz — CHANGES_REQUESTED** [review]

File: camel/retrievers/auto_retriever.py (hunk 1 above)
Comment (Wendong-Fan) [reply camel/retrievers/auto_retriever.py:55]: great point!

File: camel/retrievers/auto_retriever.py (hunk 4 above)
Comment (Wendong-Fan) [reply camel/retrievers/auto_retriever.py:161]: yeah, i have changed the docstring into `# For URLs, remove https://, replace /, and any characters not allowed by Milvus with _`

File: camel/retrievers/auto_retriever.py (hunk 6 above)
Comment (Wendong-Fan) [reply camel/retrievers/auto_retriever.py:115]: have moved some code into `_get_file_modified_date_from_storage` and `_collection_name_generator`

File: camel/retrievers/auto_retriever.py (hunk 5 above)
Comment (Wendong-Fan) [reply camel/retrievers/auto_retriever.py]: limit it to 30 characters

**Pull request merged**